src/gui.rs

```diff
@@ -394,195 +375,26 @@ impl Sam {
-                    let text = format!("{:.2}MiB / {:.2}MiB", done as f32 / (1024.0*1024.0), total as f32 / (1024.0*1024.0));
-                    ui.add_sized([ui.available_width(), 20.0], ProgressBar::new(f).text(text));
-                });
-            });
-            ui.add_space(10.0);
+        open_archive_window(ctx, &mut archive_open, self);
 
-            ui.checkbox(&mut self.replay_logs, "Replay logs");
-        });
         self.archive_window_open = archive_open;
 
         if self.data_source.simulation_settings().is_some() {
-            let old_settings = self.data_source.simulation_settings().unwrap().clone();
-
-            egui::SidePanel::left("sim").min_width(300.0).max_width(500.0).resizable(true).show(ctx, |ui| {
-                ui.set_enabled(!self.archive_window_open);
-                ui.heading("Simulation");
-                ui.add_space(20.0);
-
-                CollapsingHeader::new("Simulation Parameters").default_open(true).show(ui, |ui| {
-                    let settings = self.data_source.simulation_settings().unwrap();
-                    settings.ui(ui)
-                });
-
-                CollapsingHeader::new("(Simulated) FC Settings").default_open(false).show(ui, |ui| {
-                    let settings = self.data_source.simulation_settings().unwrap();
-                    settings.fc_settings.ui(ui, None)
-                });
-
-                ui.add_space(20.0);
-
-                let changed = *self.data_source.simulation_settings().unwrap() != old_settings;
-                let released = false;
-                ui.horizontal(|ui| {
-                    if ui.button("Reset").clicked() {
-                        *(self.data_source.simulation_settings().unwrap()) = SimulationSettings::default();
-                    }
-
-                    if ui.button("↻  Rerun").clicked() || (changed && released) {
-                        self.data_source.reset();
-                    }
-                });
-            });
+            create_simulation_panel(self, ctx);
         }
 
         // Top panel containing text indicators and flight mode buttons
-        if ctx.screen_rect().width() > 1000.0 {
-            egui::TopBottomPanel::top("topbar").min_height(60.0).max_height(60.0).show(ctx, |ui| {
-                ui.set_enabled(!self.archive_window_open);
-                ui.horizontal_centered(|ui| {
-                    self.top_bar(ui, false);
-                });
-            });
-        } else {
-            egui::TopBottomPanel::top("topbar").min_height(20.0).max_height(300.0).show(ctx, |ui| {
-                ui.set_enabled(!self.archive_window_open);
-                CollapsingHeader::new("Status & Controls").default_open(false).show(ui, |ui| {
-                    self.top_bar(ui, true);
-                    ui.add_space(10.0);
-                });
-            });
-        }
+        create_header(self, ctx);
```

Comment:
This looks good except for the same concerns about `self`. I would probably also use the opportunity to move `self.top_bar` out of here as well.